File: src/main/java/com/refinedmods/refinedstorage/item/FluidStorageDiskItem.java
package com.refinedmods.refinedstorage.item;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.RSItems;
import com.refinedmods.refinedstorage.api.storage.StorageType;
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDisk;
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskProvider;
import com.refinedmods.refinedstorage.api.storage.disk.StorageDiskSyncData;
import com.refinedmods.refinedstorage.apiimpl.API;
import com.refinedmods.refinedstorage.apiimpl.storage.FluidStorageType;
import com.refinedmods.refinedstorage.render.Styles;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.InventoryHelper;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.util.ActionResult;
import net.minecraft.util.ActionResultType;
import net.minecraft.util.Hand;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraft.world.World;
import net.minecraft.world.server.ServerWorld;

import javax.annotation.Nullable;
import java.util.List;
import java.util.UUID;

public class FluidStorageDiskItem extends Item implements IStorageDiskProvider {
    private static final String NBT_ID = "Id";

    private final FluidStorageType type;

    public FluidStorageDiskItem(FluidStorageType type) {
        super(new Item.Properties().tab(RS.MAIN_GROUP).stacksTo(1));

        this.type = type;
    }

    @Override
    public void inventoryTick(ItemStack stack, World world, Entity entity, int slot, boolean selected) {
        super.inventoryTick(stack, world, entity, slot, selected);

        if (!world.isClientSide && !stack.hasTag() && entity instanceof PlayerEntity) {
            UUID id = UUID.randomUUID();

            API.instance().getStorageDiskManager((ServerWorld) world).set(id, API.instance().createDefaultFluidDisk((ServerWorld) world, getCapacity(stack), (PlayerEntity) entity));
            API.instance().getStorageDiskManager((ServerWorld) world).markForSaving();

            setId(stack, id);
        }
    }

    @Override
    public void appendHoverText(ItemStack stack, @Nullable World world, List<ITextComponent> tooltip, ITooltipFlag flag) {
        super.appendHoverText(stack, world, tooltip, flag);

        if (isValid(stack)) {
            UUID id = getId(stack);

            API.instance().getStorageDiskSync().sendRequest(id);

            StorageDiskSyncData data = API.instance().getStorageDiskSync().getData(id);
            if (data != null) {
                if (data.getCapacity() == -1) {
                    tooltip.add(new TranslationTextComponent("misc.refinedstorage.storage.stored", API.instance().getQuantityFormatter().format(data.getStored())).setStyle(Styles.GRAY));
                } else {
                    tooltip.add(new TranslationTextComponent("misc.refinedstorage.storage.stored_capacity", API.instance().getQuantityFormatter().format(data.getStored()), API.instance().getQuantityFormatter().format(data.getCapacity())).setStyle(Styles.GRAY));
                }
            }

            if (flag.isAdvanced()) {
                tooltip.add(new StringTextComponent(id.toString()).setStyle(Styles.GRAY));
            }
        }
    }

    @Override
    public ActionResult<ItemStack> use(World world, PlayerEntity player, Hand hand) {
        ItemStack diskStack = player.getItemInHand(hand);

        if (!world.isClientSide && player.isCrouching() && type != FluidStorageType.CREATIVE) {
            IStorageDisk disk = API.instance().getStorageDiskManager((ServerWorld) world).getByStack(diskStack);

            if (disk != null && disk.getStored() == 0) {
                ItemStack storagePart = new ItemStack(FluidStoragePartItem.getByType(type), diskStack.getCount());

                if (!player.inventory.add(storagePart.copy())) {
                    InventoryHelper.dropItemStack(world, player.getX(), player.getY(), player.getZ(), storagePart);
                }

                API.instance().getStorageDiskManager((ServerWorld) world).remove(getId(diskStack));
                API.instance().getStorageDiskManager((ServerWorld) world).markForSaving();

                return new ActionResult<>(ActionResultType.SUCCESS, new ItemStack(RSItems.STORAGE_HOUSING.get()));
            }
        }

        return new ActionResult<>(ActionResultType.PASS, diskStack);
    }

    @Override
    public int getEntityLifespan(ItemStack stack, World world) {
        return Integer.MAX_VALUE;
    }

    @Override
    public UUID getId(ItemStack disk) {
        return disk.getTag().getUUID(NBT_ID);
    }

    @Override
    public void setId(ItemStack disk, UUID id) {
        disk.setTag(new CompoundNBT());
        disk.getTag().putUUID(NBT_ID, id);
    }

    @Override
    public boolean isValid(ItemStack disk) {
        return disk.hasTag() && disk.getTag().hasUUID(NBT_ID);
    }

    @Override
    public int getCapacity(ItemStack disk) {
        return type.getCapacity();
    }

    @Override
    public StorageType getType() {
        return StorageType.FLUID;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/item/FluidStorageDiskItem.kt
package com.refinedmods.refinedstorage.item

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.RSItems
import com.refinedmods.refinedstorage.api.storage.StorageType
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskProvider
import com.refinedmods.refinedstorage.apiimpl.API.Companion.instance
import com.refinedmods.refinedstorage.apiimpl.storage.FluidStorageType
import com.refinedmods.refinedstorage.render.Styles
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.InventoryHelper
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.ActionResult
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.text.Text
import net.minecraft.util.text.StringTextComponent
import net.minecraft.util.text.TranslationTextComponent
import net.minecraft.world.World
import net.minecraft.world.server.ServerWorld
import java.util.*

class FluidStorageDiskItem(type: FluidStorageType) : Item(Properties().group(RS.MAIN_GROUP).maxStackSize(1)), IStorageDiskProvider {
    private val type: FluidStorageType
    override fun inventoryTick(stack: ItemStack, world: World, entity: Entity, slot: Int, selected: Boolean) {
        super.inventoryTick(stack, world, entity, slot, selected)
        if (!world.isClient && !stack.hasTag()) {
            val id = UUID.randomUUID()
            instance().getStorageDiskManager(world as ServerWorld)!![id] = instance().createDefaultFluidDisk(world as ServerWorld, getCapacity(stack))
            instance().getStorageDiskManager(world as ServerWorld)!!.markForSaving()
            setId(stack, id)
        }
    }

    fun addInformation(stack: ItemStack?, @Nullable world: World?, tooltip: MutableList<Text?>, flag: ITooltipFlag) {
        super.addInformation(stack, world, tooltip, flag)
        if (isValid(stack)) {
            val id = getId(stack)
            instance().getStorageDiskSync()!!.sendRequest(id)
            val data = instance().getStorageDiskSync()!!.getData(id)
            if (data != null) {
                if (data.getCapacity() == -1) {
                    tooltip.add(TranslationTextComponent("misc.refinedstorage.storage.stored", instance().getQuantityFormatter()!!.format(data.getStored())).setStyle(Styles.GRAY))
                } else {
                    tooltip.add(TranslationTextComponent("misc.refinedstorage.storage.stored_capacity", instance().getQuantityFormatter()!!.format(data.getStored()), instance().getQuantityFormatter()!!.format(data.getCapacity())).setStyle(Styles.GRAY))
                }
            }
            if (flag.isAdvanced()) {
                tooltip.add(StringTextComponent(id.toString()).setStyle(Styles.GRAY))
            }
        }
    }

    fun onItemRightClick(world: World, player: PlayerEntity, hand: Hand?): ActionResult<ItemStack> {
        val diskStack: ItemStack = player.getHeldItem(hand)
        if (!world.isClient && player.isCrouching() && type !== FluidStorageType.CREATIVE) {
            val disk = instance().getStorageDiskManager(world as ServerWorld)!!.getByStack(diskStack)
            if (disk != null && disk.getStored() == 0) {
                val storagePart = ItemStack(FluidStoragePartItem.Companion.getByType(type), diskStack.count)
                if (!player.inventory.addItemStackToInventory(storagePart.copy())) {
                    InventoryHelper.spawnItemStack(world, player.getPosX(), player.getPosY(), player.getPosZ(), storagePart)
                }
                instance().getStorageDiskManager(world as ServerWorld)!!.remove(getId(diskStack))
                instance().getStorageDiskManager(world as ServerWorld)!!.markForSaving()
                return ActionResult(ActionResult.SUCCESS, ItemStack(RSItems.STORAGE_HOUSING))
            }
        }
        return ActionResult(ActionResult.PASS, diskStack)
    }

    fun getEntityLifespan(stack: ItemStack?, world: World?): Int {
        return Int.MAX_VALUE
    }

    override fun getId(disk: ItemStack?): UUID? {
        return disk!!.tag.getUniqueId(NBT_ID)
    }

    override fun setId(disk: ItemStack?, id: UUID?) {
        disk!!.tag = CompoundTag()
        disk.tag.putUniqueId(NBT_ID, id)
    }

    override fun isValid(disk: ItemStack?): Boolean {
        return disk!!.hasTag() && disk.tag.hasUniqueId(NBT_ID)
    }

    override fun getCapacity(disk: ItemStack?): Int {
        return type.getCapacity()
    }

    override fun getType(): StorageType? {
        return StorageType.FLUID
    }

    companion object {
        private const val NBT_ID = "Id"
    }

    init {
        this.type = type
        this.setRegistryName(RS.ID, type.getName().toString() + "_fluid_storage_disk")
    }
}